### PR TITLE
feat: add resourcetopo package

### PR DESCRIPTION
**Pull request by AnnaYue**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [x] N
- [ ] Y 

#### 2. What is the scope of this PR (e.g. component or file name):
resourcetopo/

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [x] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Other

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

#### 6. Release note

Please refer to [Release Notes Language Style Guide](https://kusionstack.io/docs/governance/release-policy/) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: resourcetopo/manager.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+}
+func (m *manager) AddObjectHandler(meta metav1.TypeMeta, handler NodeHandler) error {
+	s := m.storages[generateMetaKey(&meta)]
+	if s == nil {
+		return fmt.Errorf("resource %v not configured in this manager", meta)
+	}
+	return s.AddObjectHandler(handler)
+}
+
+func (m *manager) GetTopo(meta metav1.TypeMeta, name types.NamespacedName) (NodeInfo, error) {
```
Comment (shaofan-hs) [inline resourcetopo/manager.go:95]: How about GetNodeInfo?

File: resourcetopo/manager.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+	return s, nil
+}
+
+func (m *manager) getStorage(meta metav1.TypeMeta) *nodeStorage {
+	key := generateMetaKey(&meta)
+
+	return m.storages[key]
+}
+
+func (m *manager) createVirtualStorage(object metav1.TypeMeta) (*nodeStorage, error) {
```
Comment (shaofan-hs) [inline resourcetopo/manager.go:131]: Change object to meta, or keep them consistent?

File: resourcetopo/dag_util.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+package resourcetopo
+
+import (
+	"container/list"
+	"fmt"
+)
+
+// dagCheck implement a depth first algorithm to make sure the resource topology in this manager has no cycle
+func (m *manager) dagCheck() error {
```
Comment (shaofan-hs) [inline resourcetopo/dag_util.go:25]: We can merge these util go files to only one, and move this method to manager.go

File: resourcetopo/node_info.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+	labels    labels.Set // labels is a ref to object.meta.labels, do not edit!
+	relations []ResourceRelation
+}
+
+func newNode(s *nodeStorage, namespace, name string) *nodeInfo {
+	return &nodeInfo{
+		storageRef: s,
+		name:       name,
+		namespace:  namespace,
+		fakeNode:   !s.virtualResource,
```
Comment (shaofan-hs) [inline resourcetopo/node_info.go:53]: VirtualNode is easier to understand.

File: resourcetopo/node_info.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+	}
+}
+
+func (n *nodeInfo) readyToDelete() bool {
+	n.lock.RLock()
+	defer n.lock.RUnlock()
+	return n.fakeNode && (n.directReferredPreOrders == nil || n.directReferredPreOrders.Len() == 0)
+}
+
+func (n *nodeInfo) isFakeNode() bool {
```
Comment (shaofan-hs) [inline resourcetopo/node_info.go:171]: isVirtualNode?

File: resourcetopo/manager.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+func (m *manager) getOrCreateStorage(typeMeta metav1.TypeMeta, getInformer func(meta metav1.TypeMeta) cache.SharedInformer) (*nodeStorage, error) {
+	key := generateMetaKey(&typeMeta)
+
+	s, ok := m.storages[key]
+	if ok {
+		return s, nil
+	}
+
+	informer := getInformer(typeMeta)
```
Comment (shaofan-hs) [inline resourcetopo/manager.go:115]: Check whether `getInfomer` is nil?

File: resourcetopo/node_storage.go [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+	defer s.handlersLock.Unlock()
+	if s.relationUpdateHandler == nil {
+		s.relationUpdateHandler = make(map[string][]RelationHandler)
+	}
+	postKey := generateMetaKey(postMeta)
+
+	handlers := s.relationUpdateHandler[postKey]
+	for _, h := range handlers {
+		if h == relationHandler {
+			return fmt.Errorf("relation handler %v already added", relationHandler)
```
Comment (shaofan-hs) [inline resourcetopo/node_storage.go:103]: How about `return nil`?

File: resourcetopo/node_storage.go [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+	return &nodeStorage{
+		manager:         manager,
+		meta:            meta,
+		metaKey:         generateMetaKey(&meta),
+		virtualResource: true,
+		namespacedInfo:  make(map[string]map[string]*nodeInfo),
+	}
+}
+
+func (s *nodeStorage) addRelationHandler(postMeta *metav1.TypeMeta, relationHandler RelationHandler) error {
```
Comment (shaofan-hs) [inline resourcetopo/node_storage.go:92]: There is no need to return error?

File: resourcetopo/manager.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+	if ok {
+		return nil, fmt.Errorf("unexpected twice configuration for virtual resource %s", key)
+	}
+
+	s := newVirtualStorage(m, object)
+	m.storages[key] = s
+	return s, nil
+}
+
+func (m *manager) AddRelationHandler(preOrder, postOrder metav1.TypeMeta, handler RelationHandler) error {
```
Comment (shaofan-hs) [inline resourcetopo/manager.go:143]: We can add some comments for those key methods.

File: resourcetopo/manager.go [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+	}
+
+	if err := m.AddConfig(cfg); err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+func (m *manager) AddConfig(cfg Config) error {
```
Comment (shaofan-hs) [inline resourcetopo/manager.go:53]: addConfig may be more safety.

File: resourcetopo/util.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func generateMetaKey(meta *metav1.TypeMeta) string {
+	return fmt.Sprintf("%s/%s", meta.APIVersion, meta.Kind)
+}
+
+func compareResourceRelation(a, b ResourceRelation) int {
```
Comment (shaofan-hs) [inline resourcetopo/util.go:31]: Add some tests for this utils funcs?

File: resourcetopo/node_topology_updater.go [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+		for _, preStorage := range s.preOrderResources {
+			preStorage.checkForLabelUpdate(node)
+		}
+	}
+
+	for _, discover := range s.discoverers {
+		newDiscoverd := discover.Discover(newTopoObj)
+		oldDiscoverd := discover.Discover(oldTopoObj)
+		slices.SortFunc(newDiscoverd, compareObjectMeta)
+		slices.SortFunc(newDiscoverd, compareObjectMeta)
```
Comment (shaofan-hs) [inline resourcetopo/node_topology_updater.go:92]: slices.SortFunc(oldDiscoverd, compareObjectMeta)?

File: resourcetopo/node_topology_updater.go [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+
+var _ cache.ResourceEventHandler = &nodeStorage{}
+
+func (s *nodeStorage) OnAdd(obj interface{}) {
+	topoObject, ok := obj.(Object)
+	if !ok {
+		klog.Errorf("Failed to transform to k8s object %v, ignore this update nodeEvent", obj)
+		return
+	}
+	node := s.getNode(topoObject.GetNamespace(), topoObject.GetName())
```
Comment (shaofan-hs) [inline resourcetopo/node_topology_updater.go:36]: Use getOrCreateNode instead?

File: resourcetopo/manager.go (hunk 6 above)
Comment (AnnaYue) [reply resourcetopo/manager.go:141]: add nil check in manager call func

File: resourcetopo/manager.go (hunk 9 above)
Comment (AnnaYue) [reply resourcetopo/manager.go:169]: added in types interface define

File: resourcetopo/node_storage.go [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+
+	return node
+}
+
+func (s *nodeStorage) deleteNode(namespace, name string) {
+	if isClusterNamespace(namespace) {
+		namespace = allNamespaceKey
+	}
+
+	namedInfo := s.namespacedInfo[namespace]
```
[the remark this answers is not in the document]
Comment (AnnaYue) [reply resourcetopo/node_storage.go:194]: done

File: resourcetopo/node_storage.go [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+	return &nodeStorage{
+		manager:         manager,
+		meta:            meta,
+		metaKey:         generateMetaKey(&meta),
+		virtualResource: true,
+		namespacedInfo:  make(map[string]map[string]*nodeInfo),
+	}
+}
+
+func (s *nodeStorage) GetNode(namespacedName *types.NamespacedName) (NodeInfo, error) {
```
Comment (shaofan-hs) [inline resourcetopo/node_storage.go]: When use namespace and name, and when use NamespacedName?

File: resourcetopo/resourcetopo_utils_test.go [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+	"github.com/hashicorp/consul/sdk/testutil/retry"
+	. "github.com/onsi/ginkgo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+var (
+	ReplicaSetMeta = metav1.TypeMeta{Kind: "ReplicaSet", APIVersion: "apps/v1"}
```
Comment (shaofan-hs) [inline resourcetopo/resourcetopo_utils_test.go:31]: Use name or short name in the same?

File: resourcetopo/resourcetopo_base_test.go [hunk 17]
```diff
@@ -0,0 +1,514 @@
...
+	})
+	AfterEach(func() {
+		if !checkAll() {
+			klog.Infof("end with object [sts %s, pod %s, ists %s]", stsHandler.string(), podHandler.string(), istsHandler.string())
+			klog.Infof("end with relation [stsIsts %s, podSts %s]", stsIstsRelation.string(), podStsRelation.string())
+		}
+		cancel()
+	})
+
+	It("single pod add", func() {
```
Comment (shaofan-hs) [inline resourcetopo/resourcetopo_base_test.go:90]: `Single pod added` or `Add single pod`?

File: resourcetopo/resourcetopo_base_test.go [hunk 18]
```diff
@@ -0,0 +1,514 @@
...
+	return &TopologyConfig{
+		GetInformer: func(meta metav1.TypeMeta) cache.SharedInformer {
+			return GetInformer(meta, k8sInformerFactory)
+		},
+		Resolver: []RelationResolver{
+			{
+				PreMeta:   StsMeta,
+				PostMetas: []metav1.TypeMeta{PodMeta},
+				Resolve: func(preOrder Object) []ResourceRelation {
+					stsObj := preOrder.(*appsv1.StatefulSet)
```
Comment (shaofan-hs) [inline resourcetopo/resourcetopo_base_test.go:443]: Check `ok` may be safer

**Review by shaofan-hs — APPROVED** [review]

**Pull request merged**